var/spack/repos/builtin/packages/upcxx/package.py

```diff
@@ -64,26 +49,29 @@ def setup_environment(self, spack_env, run_env):
         if 'platform=cray' in self.spec:
             spack_env.set('GASNET_CONFIGURE_ARGS', '--enable-mpi=probe')
             if "scheduler=slurm" in self.spec:
-                tty.warn("CROSS has been automatically set to %s."
-                         % 'cray-aries-slurm')
                 spack_env.set('CROSS', 'cray-aries-slurm')
             elif "scheduler=alps" in self.spec:
-                tty.warn("CROSS has been automatically set to %s."
-                         % 'cray-aries-alps')
                 spack_env.set('CROSS', 'cray-aries-alps')
         if '+cuda' in self.spec:
             spack_env.set('UPCXX_CUDA', '1')
             spack_env.set('UPCXX_CUDA_NVCC', self.spec['cuda'].prefix.bin.nvcc)
 
+        run_env.set('UPCXX_INSTALL', self.prefix)
+        run_env.set('UPCXX', self.prefix.bin.upcxx)
+        if 'platform=cray' in self.spec:
+            run_env.set('UPCXX_GASNET_CONDUIT', 'aries')
+            run_env.set('UPCXX_NETWORK', 'aries')
+
     def setup_dependent_package(self, module, dep_spec):
         dep_spec.upcxx = self.prefix.bin.upcxx
 
     def setup_dependent_environment(self, spack_env, run_env, dependent_spec):
-        spack_env.set('UPCXX_INSTALL', self.prefix)
-        spack_env.set('UPCXX', self.prefix.bin.upcxx)
         if 'platform=cray' in self.spec:
             spack_env.set('UPCXX_GASNET_CONDUIT', 'aries')
+            spack_env.set('UPCXX_NETWORK', 'aries')
 
     def install(self, spec, prefix):
-        installsh = Executable('./install')
+        env['CC'] = self.compiler.cc
```

Comment:
@mjacquel Spack should set `CC` and `CXX` automatically: are you seeing an error when you don't set them?

Reply:
@scheibelp , the issue is the following: if I do not set them, I do not get an error at build-time nor when using UPC++ in other spack packages (by using `depends`). However, UPC++'s installer picks up spack's compiler wrapper as its own internal compiler, and when a user uses `spack load upcxx`, UPC++'s compiler wrapper (named `upcxx`) will try to use Spack's compiler wrapper, which then complains about all `SPACK_*` environment variables not being set. 

One workaround is to do `spack build-env upcxx upcxx -O3 -o hello-world hello-world.cpp`, which is not clean, and very disturbing for the end-user.

Setting `CC` to the full path does allow UPC++'s `upcxx` wrapper to directly call the actual compiler.

Is there any better solution to do this?

Reply:
@scheibelp Spack sets `CC` to `spack_cc` (the compiler wrapper), this PR instead sets it to `self.compiler.cc` (the compiler). Based on:

> fixes an issue with the compiler passed on to the UPC++ install script (to get the full path to the actual compiler instead of Spack's compiler wrapper) in order to achieve the correct behavior when doing `spack load upcxx`.

I'm going to assume that UPC++ tries to use the Spack compiler wrapper at run-time, and this change is necessary to prevent that. If only there was a text file we could patch at run-time to replace the compiler wrapper with the actual compiler.

Reply:
@mjacquel You beat me to it!

> Is there any better solution to do this?

If UPC++'s `upcxx` is a text file that contains the compiler wrapper, you can patch it after installation. If not, your current solution is the only workaround.

Reply:
Mmmm I think there is one. I need to do some digging. What is the correct procedure for applying post-install patches?

Reply:
In that case, drop these two lines, then add something like the following at the end:
```python
filter_file(spack_cc, self.compiler.cc, prefix.bin.upcc)
filter_file(spack_cxx, self.compiler.cxx, prefix.bin.upcxx)
```

Reply:
> > Is there any better solution to do this?
> 
> If UPC++'s `upcxx` is a text file that contains the compiler wrapper, you can patch it after installation. If not, your current solution is the only workaround.

This solution would **NOT** be a good one for UPC++. 

The install process embeds the backend compiler path into over a dozen files in the install tree. `upcxx` is a script that reads the compiler variable and other settings from one of up to (currently) ten different text files. It's also embedded into the GASNet makefile fragments that get installed along with UPC++. Changing the compiler path post-install is something we never test or support for end users and would make for a very brittle spack package.

Reply:
So, clearly we shouldn't do a post-install patch. Then, is there any particular reason why we shouldn't do what we are doing right now in the spack package?

Reply:
Also, I believe I took the exact same approach than the OpenMPI spack package.

Reply:
>  is there any particular reason why we shouldn't do what we are doing right now in the spack package?

Spack's compiler wrappers are convenient because they automatically inject things like `-I/path/to/include`, `-L/path/to/lib`, and `-Wl,rpath,/path/to/lib`, among other things. Since you only have a single dependency, the downside of your current solution is that it may not be able to find cuda at build/run time depending on the system. Personally, if I were you, I would patch all files containing the compiler string post-installation. We do this in several other packages as well (see `filter_compilers` in `cantera`, `psi4`, `python`, and `sundials`)

Reply:
With that said, if you want to use your current solution and deal with the possibility that dependencies might not be RPATHed, I'm fine with that too.

Reply:
The `upcxx` compiler wrapper is exactly analogous to the `mpicc` and `mpicxx` compiler wrappers used by OpenMPI and MPICH. As with those wrappers, it needs direct access to the underlying compilers long after Spack has finished installing the software, and very often from other users (who may or may not have Spack in their path). I'll admit I'm not that familiar with Spack internals, but Mathias followed the example of the established OpenMPI package that is presumably doing it in a manner that is correct, or at least acceptable.

It's also worth noting that the UPC++ install process already includes significant intelligence to extract the CUDA install path information from `nvcc` and add it to `upcxx`, because that's needed anyhow to ensure correct operation on manual non-spack installs.

Reply:
> Mathias followed the example of the established OpenMPI package that is presumably doing it in a manner that is correct, or at least acceptable.

The `openmpi` package actually does exactly what I'm suggesting. It calls `filter_compiler_wrappers` (like `hdf5`, `mpich`, `mvapich2`, `neuron`, `petsc`, `r,` and `tau`) to replace the Spack compiler wrappers with the actual compiler post-installation. It supports recursively replacing files, so it should be pretty stable.

Reply:
See https://spack.readthedocs.io/en/latest/spack.html#spack.mixins.filter_compiler_wrappers for docs

Reply:
@adamjstewart I think I got confused with what I did. I have a question regarding `filter_compiler_wrappers`. Does it goes through binary files as well if I provide a base directory and ask for a recursive change?

Reply:
It does not handle binary files, only text files. You can see the definition in `lib/spack/spack/mixins.py`. It's a wrapper around the `filter_file` and `FileFilter` methods defined in `lib/spack/llnl/util/filesystem.py`.

Reply:
So it should be safe to call on a directory then?

Reply:
Yep, although it still expects a list of file names (which can be found recursively if necessary)

Reply:
> Spack's compiler wrappers are convenient because they automatically inject things like `-I/path/to/include`, `-L/path/to/lib`, and `-Wl,rpath,/path/to/lib`, among other things. [...] Personally, if I were you, I would patch all files containing the compiler string post-installation. We do this in several other packages as well (see `filter_compilers` in `cantera`, `psi4`, `python`, and `sundials`)

@adamjstewart : Assuming we can get this `filter_compiler_wrappers` hack to work, I still don't see how this helps specifically with the CUDA RPATH as you describe. All our linkage happens at end-user compile time, long after spack is done running. As I understand it, you are advocating we build/install UPC++ using the wrapper, then "swap" the linker invoked by our compiler wrapper from the spack wrapper to instead be the raw backend compiler. But I think that change loses the insertion of the exact rpath flags you are arguing as motivation - when the end user invokes `upcxx` they will get the raw compiler linker without any flags the Spack wrapper might have added.

What am I missing?

Reply:
@bonachea Yes, the compiler wrappers are only effective at Spack compile time, not at run-time. Is there no linking involved during the `upcxx` build? If that's the case, it doesn't make much of a difference. Although if you plan on adding packages that build with `upcxx` to Spack, you can still use Spack's compiler wrappers. This is why `openmpi` sets `OMPI_CC` and friends, for example.

Reply:
> @bonachea Yes, the compiler wrappers are only effective at Spack compile time, not at run-time. Is there no linking involved during the `upcxx` build? If that's the case, it doesn't make much of a difference. 

There's essentially no linking during `upcxx` build - the primary product is a set of static libraries and associated headers and scripts. There are a few utility executables that might optionally get linked, but none of them rely on CUDA or other external dependencies.

> Although if you plan on adding packages that build with `upcxx` to Spack, you can still use Spack's compiler wrappers. This is why `openmpi` sets `OMPI_CC` and friends, for example.

There are currently no such packages. I guess they could exist in the future, but without a use case this seems like we are motivating adding complexity with no currently measurable benefit...

Reply:
If you're fine with skipping the compiler wrappers entirely and setting `CC` and `CXX`, I'm fine with that.

Reply:
I agree with 

> If you're fine with skipping the compiler wrappers entirely

but I'd also like to add a note on:

> Does it goes through binary files as well if I provide a base directory and ask for a recursive change?

If there are instances of the compiler embedded in binaries I think that is another reason to skip using `filter_compiler_wrappers` (since the outcome would be inconsistent) and stick with setting `CC`/`CXX` to the underlying compiler executables. Adam's points about what is lost still stand but may be mitigated here by

* setting an environment variable related to the only dependency (cuda): UPCXX_CUDA_NVCC
* using a module (e.g. a Spack-generated module) which updated LD_LIBRARY_PATH

It seems for this particular build, we would have to add support in Spack to allow building with the Spack compiler wrappers and also allowing use of `upcxx` afterwards.

Reply:
> If you're fine with skipping the compiler wrappers entirely and setting CC and CXX, I'm fine with that.

I think this is fine for now, given that it's not really possible to filter the compilers post-build.  It would be really nice, though, if `upcxx` could support this in the future.

> Spack's compiler wrappers are convenient because they automatically inject things like `-I/path/to/include`, `-L/path/to/lib`, and `-Wl,rpath,/path/to/lib`, among other things.

Among the more recent other things are the specific target arguments (see [here]((https://spack.readthedocs.io/en/latest/basic_usage.html#architecture-specifiers), [here](https://spack.readthedocs.io/en/latest/packaging_guide.html#querying-the-target-microarchitecture)) now that we'll detect and build for specific microarchitectures.  This is part of our effort to make it so that we can actually distribute optimized binaries with Spack.

If `upcxx` doesn't build with the compiler wrappers, it'll be up to the `upcxx` build to inject the right flags.  If users try any of these:

```console
$ spack install upcxx cxxflags="-O3 -fast"
$ spack install upcxx target=skylake_avx512
```

They'll be ignored unless the package does something about it.

Anyway, let's just let it build with the backend compiler for now.  It is good to just have `upcxx` in Spack in the first place.

To support the nicer compiler wrapper features, it seems like the `upcxx` folks would just need to fix this (from @bonachea above):

> The install process embeds the backend compiler path into over a dozen files in the install tree. upcxx is a script that reads the compiler variable and other settings from one of up to (currently) ten different text files.

and they would probably want to make it so that the user can inject a compiler in the build, via a variable like `OMPI_CC` or `MPICH_CC`, like most MPI's do.  That would make it easier for an ecosystem of packages *using* `upcxx` to grow up in Spack.

Reply:
I propose we move on with the current approach, since compatibility with `spack load` was something we have been requested to provide. Let me know if any of you disagree.